receiver/snmpreceiver/scraper.go

```diff
@@ -39,13 +90,489 @@ func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCrea
+// into metrics with optional attribute and/or resource attributes
+func (s *snmpScraper) scrapeIndexedMetrics(
+	now pcommon.Timestamp,
+	resourceMetricsMap map[string]*pmetric.ResourceMetrics,
+	metricsByResource *metricsByResourceContainer,
+	scraperErrors *scrapererror.ScrapeErrors,
+) {
+	// Retrieve column OID SNMP indexed data for attributes
+	indexedAttributeValues := map[indexedAttributeKey]string{}
+	s.scrapeIndexedAttributes(indexedAttributeValues, scraperErrors)
+
+	// Retrieve column OID SNMP indexed data for resource attributes
+	indexedResourceAttributeValues := map[indexedAttributeKey]string{}
+	s.scrapeIndexedResourceAttributes(indexedResourceAttributeValues, scraperErrors)
+
+	// Find all metric column OIDs
+	// Also create a map of metric names with OID as key so the metric config will be easy to
+	// matchup later with returned SNMP indexed data
+	indexedMetricNamesByOID := map[string]string{}
+	indexedMetricOIDs := []string{}
+	for name, metricCfg := range s.cfg.Metrics {
+		if len(metricCfg.ColumnOIDs) > 0 {
+			for i, oid := range metricCfg.ColumnOIDs {
+				// Data is returned by the client with '.' prefix on the OIDs.
+				// Making sure the prefix exists here in the configs so we can match it up with returned data later
+				if !strings.HasPrefix(oid.OID, ".") {
+					oid.OID = "." + oid.OID
+					s.cfg.Metrics[name].ColumnOIDs[i].OID = oid.OID
+				}
+				indexedMetricOIDs = append(indexedMetricOIDs, oid.OID)
+				indexedMetricNamesByOID[oid.OID] = name
+			}
+		}
+	}
+
+	// If no column metric configs, nothing else to do
+	if len(indexedMetricOIDs) == 0 {
+		return
+	}
+
+	// Get all column OID SNMP indexed data for metrics and turn it into metrics, attributes,
+	// and resource attributes (using the previously retrieved attribute and resource attribute data)
+	s.client.GetIndexedData(
+		indexedMetricOIDs,
+		indexedDataToMetric(
+			now, resourceMetricsMap, metricsByResource, indexedMetricNamesByOID,
+			indexedAttributeValues, indexedResourceAttributeValues, s,
+		),
+		scraperErrors,
+	)
+}
+
+// indexedDataToMetric provides a function which will convert one piece of column OID SNMP indexed data
+// and turn it into a metric with attributes based on the config and previously collected column OID
+// SNMP indexed attribute. A resource may also be created if none exists for the related previously
+// collected resource attribute data yet.
+func indexedDataToMetric(
+	now pcommon.Timestamp,
+	resourceMetricsMap map[string]*pmetric.ResourceMetrics,
+	metricsByResource *metricsByResourceContainer,
+	indexedMetricNamesByOID map[string]string,
+	indexedAttributeValues map[indexedAttributeKey]string,
+	indexedResourceAttributeValues map[indexedAttributeKey]string,
+	snmpScraper *snmpScraper,
+) processFunc {
+	// This returns a processFunc because this is what the client's GetScalarData method requires
+	return func(data snmpData) error {
```

Comment:
I'm having a really hard time understanding this function. There are so many maps being passed around and so many different things being done in this one function. Can you find a way to organize the work into smaller, clearer steps?

Reply:
@djaglowski i'll break this one up a bit